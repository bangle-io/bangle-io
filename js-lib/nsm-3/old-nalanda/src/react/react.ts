import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
