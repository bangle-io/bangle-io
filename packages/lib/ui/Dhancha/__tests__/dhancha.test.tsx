/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { render } from '@testing-library/react';
import React from 'react';

import { DhanchaWidescreen } from '../index';

describe('DhanchaWidescreen component', () => {
  it('renders correctly and handles resizing', () => {
    const onChangeWidthMock = jest.fn();
    const { container } = render(
      <DhanchaWidescreen
        activitybar={<div>Activity Bar</div>}
        leftAside={<div>Left Aside</div>}
        mainContent={<div>Main Content</div>}
        rightAside={<div>Right Aside</div>}
        titlebar={<div>Title Bar</div>}
        onChangeWidth={onChangeWidthMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
