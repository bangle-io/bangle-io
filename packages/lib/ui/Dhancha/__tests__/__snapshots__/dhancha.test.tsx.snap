// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DhanchaWidescreen component renders correctly and handles resizing 1`] = `
<div>
  <div
    class="B-ui-dhancha-widescreen B-ui-has-left-aside B-ui-has-right-aside B-ui-has-activitybar "
  >
    <div
      aria-label="Title Bar"
      class="B-ui-activitybar"
      role="navigation"
      style="grid-area: activitybar;"
    >
      <div>
        Activity Bar
      </div>
    </div>
    <aside
      class="B-ui-left-aside"
      style="grid-area: left-aside;"
    >
      <div>
        Left Aside
      </div>
    </aside>
    <div
      class="B-ui-aside-separator "
      style="grid-area: left-aside-separator;"
    >
      <div />
      <div />
    </div>
    <div
      aria-label="Title Bar"
      class="B-ui-dhancha-titlebar titlebar-container"
      role="navigation"
      style="grid-area: titlebar-container;"
    >
      <div>
        Title Bar
      </div>
    </div>
    <div
      class="B-ui-main-content"
      style="grid-area: main-container;"
    >
      <div>
        Main Content
      </div>
    </div>
    <div
      class="B-ui-aside-separator "
      style="grid-area: right-aside-separator;"
    >
      <div />
      <div />
    </div>
    <aside
      class="B-ui-right-aside"
      style="grid-area: right-aside;"
    >
      <div>
        Right Aside
      </div>
    </aside>
  </div>
</div>
`;
